cmd/fleet/handleEnroll.go

```diff
@@ -238,11 +238,96 @@ func _enroll(ctx context.Context, bulker bulk.Bulk, c cache.Cache, req EnrollReq
 	}
 
 	// We are Kool & and the Gang; cache the access key to avoid the roundtrip on impending checkin
-	c.SetApiKey(*accessApiKey, true)
+	et.cache.SetApiKey(*accessApiKey, true)
 
 	return &resp, nil
 }
 
+// Remove the ghost artifacts from Elastic; the agent record and the accessApiKey.
+func (et *EnrollerT) wipeGhosts(ctx context.Context, zlog zerolog.Logger, resp *EnrollResponse) {
+	zlog = zlog.With().Str(LogAgentId, resp.Item.ID).Logger()
+
+	if err := et.bulker.Delete(ctx, dl.FleetAgents, resp.Item.ID); err != nil {
+		zlog.Error().Err(err).Msg("ghost agent record failed to delete")
+	} else {
+		zlog.Info().Msg("ghost agent record deleted")
+	}
+
+	invalidateApiKey(ctx, zlog, et.bulker, resp.Item.AccessApiKeyId)
+}
+
+func invalidateApiKey(ctx context.Context, zlog zerolog.Logger, bulker bulk.Bulk, apikeyId string) error {
```

Comment:
this has a potential to hold the request go routine alive, that might cause the go routines exhaustion at scale.
agreed this will be addressed in the follow up PRs probably as standalone GC routine/pool

Reply:
Agreed, will address in subsequent PR.

Reply:
Is there already a follow up PR for this?

Reply:
not yet, was planning to address that